.github/workflows/haskell-linux.yml

```diff
@@ -158,6 +158,22 @@ jobs:
         KEEP_WORKSPACE: 1
       run: cabal test cardano-testnet cardano-api cardano-node cardano-node-chairman cardano-cli cardano-submit-api
 
+    - name: Tar failed tests workspaces
+      if: ${{ failure() }}
+      env:
+        TMP: ${{ runner.temp }}
+      shell: bash
+      run: |
+        cd $TMP
+        find . -name 'module' -type f -printf '%h\n' | xargs -L1 basename | sort -u | xargs tar -czvf workspaces.tgz
+
+    - name: Upload workspaces on tests failure
+      if: ${{ failure() }}
+      uses: actions/upload-artifact@v3
+      with:
+        name: failed-test-workspaces-${{ matrix.os }}-ghc${{ matrix.ghc }}.tgz
```

Comment:
This will be overwritten for every cabal version in the job matrix (might cause failures in case of concurrent uploads). Currently there's a single version, so it should be fine, hence I skipped the cabal version in the file name. Not sure if it's worth including here.

Reply:
I think it's worth it to prevent confusion in the case of multiple failures.